scalyr_agent/agent_main.py

```diff
@@ -1069,7 +1069,527 @@ def __get_log_files_initial_positions(self, only_new_files=False):
+                    if time.time() >= last_essential_monitors_check_time + 5:
+                        self._fail_if_essential_monitors_are_not_running()
+                        last_essential_monitors_check_time = time.time()
+
+                # The while loop is over, meaning that the agent is about to stop.
+                # Before fully exiting, we have to be sure that the config check generator has also finished and
+                # performed all its cleanups. Since the generator also looks at the '__run_state' to know when
+                # to finish, we just advance it once more with stopped run state.
+                try:
+                    next(check_config_for_change_gen)
+                except StopIteration:
+                    pass
+
+            finally:
+                # If there is an error from the outside of the check config generator, or the generator itself
+                # hasn't been exhausted for this moment, then close it, so it can do its cleanups.
+                # NOTE: the 'close' method is just ignored if generator is already exhausted.
+                check_config_for_change_gen.close()
+
+        except Exception:
+            log.exception(
+                "Main run method for agent failed due to exception",
+                error_code="failedAgentMain",
+            )
+        finally:
+            # NOTE: We manually call close_handlers() here instead of registering it to call it as
+            # part of run state stop routine.
+            # The reason for that is that run state stop callbacks are called before we get here
+            # which means that some messages which are produced after that and before fully shutting
+            # down are lost and not logged.
+            scalyr_logging.close_handlers()
+
+    def _fail_if_essential_monitors_are_not_running(self):
+        """
+        Searches for dead monitors in monitor manager's status and raises
+            error if dead "essential" monitors are found.
+
+        Essential means that the monitor has a special configuration option which tells
+        agent that it can not continue operating without this monitor.
+        """
+        monitor_manager_status = self.__monitors_manager.generate_status()
+        for monitor_status in monitor_manager_status.monitors_status:
+            if monitor_status.is_alive:
+                continue
+
+            stopped_monitor = self.__monitors_manager.find_monitor_by_short_hash(
+                monitor_status.monitor_short_hash
+            )
+            if stopped_monitor.stop_agent_on_failure:
+                raise Exception(
+                    "Monitor '{}' with short hash '{}' is not running, stopping the agent because it is configured "
+                    "not to run without this monitor.".format(
+                        monitor_status.monitor_name, monitor_status.monitor_short_hash
+                    )
+                )
+
     def __run(self, controller):
```

Comment:
So you will remove this method just before merging and rename ``__new_run`` to ``__run``?

Reply:
Yes, removed, thanks.